### ISSUE TITLE
multi cpu training is worse than single cpu

### ISSUE THREAD

**Issue by mariamrahmaani** [horovod/horovod#1419]
**Environment:**
1. Framework: (TensorFlow, Keras, PyTorch, MXNet)      Pytorch
2. Framework version:  latest code from huggingface: https://github.com/huggingface/pytorch-transformers 

3. Horovod version: 0.18.1
4. MPI version: 4.0.1
5. CUDA version: None
6. NCCL version: None
7. Python version: Python 3.6.8
8. OS and version: ubuntu 18.04.2 LTS
9. GCC version: 7.4.0

**Checklist:**
1. Did you search issues to find if somebody asked this question before? yes
2. If your question is about hang, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/running.rst)?
3. If your question is about docker, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/docker.rst)?
4. Did you check if you question is answered in the [troubleshooting guide](https://github.com/horovod/horovod/blob/master/docs/troubleshooting.rst)? Yes

**Bug report:**
Please describe errorneous behavior you're observing and steps to reproduce it.

I am using Huggingface Bert pytorch implementation to fine-tune on 2 servers just using CPUs.

When I am not using Horovod the finetuning task (dataset:MRPC + Model: BertBase+batchSize: 64 + Epoch: 3 + other default setting ) on one CPU server will give me Accuracy=0.813

When I use 2 servers with Horovod code added to the experiment with the same setup I get Accuracy=0.6887

Here is the script that I am using:
https://github.com/mariamrahmaani/pytorch-transformers/blob/master/pytorch-transformers/examples/run_glue_hvd.py

https://github.com/mariamrahmaani/pytorch-transformers/blob/master/pytorch-transformers/examples/horovod.sh

I tried couple of experiments but none of these seems to work for accuracy more than 0.6887 :
 -  used lr=args.learning_rate   instead of    lr = lr=args.learning_rate * hvd.size() 
 - Weight_Decay = 0.1   instead of    0.0
 - Higher number of Epoch (10 instead of 3)
 - smaller batch size (32 instead of 64)
- used higher number of threads per server (2 for each server instead of 1): np 4 -H ip1:2,ip2:2

Can you please look at the code and direct me if I am doing something wrong in the implementation side of Horovod.

Thanks,
Mariam

**Comment by mariamrahmaani** [comment]
Any idea?

**Comment by abditag2** [comment]
When you use horovod, you need to scale down your batch size by the hvd.size() so that your effective batch size across all the workers becomes equal to when not using Horovod. In other words, in your code, set the batch size to bath_size.hvd.size()

Not sure if this is exactly related to the accuracy you are seeing.

**Comment by mariamrahmaani** [comment]
I made sure to scale down the batch size, and  I tried smaller batch size and still I get no improvement in accuracy. I used size of 32 with 2 CPU (hvd.size() = 2) which should be equal to using Batch size=64 without using horovod. My total time to train improves near 100% but accuracy as I said decreases 13%.

Thanks for any other help/idea you may have.

**Comment by zira-wang** [comment]
> I made sure to scale down the batch size, and I tried smaller batch size and still I get no improvement in accuracy. I used size of 32 with 2 CPU (hvd.size() = 2) which should be equal to using Batch size=64 without using horovod. My total time to train improves near 100% but accuracy as I said decreases 13%.
> 
> Thanks for any other help/idea you may have.

Have you solved the problem ？I have the same problem .

**Comment by mariamrahmaani** [comment]
Not resolved yet

**Comment by tgaddair** [comment]
Hey @mariamrahmaani and @zira-wang, what I would suggest doing is first is to try running one of our [example](https://github.com/horovod/horovod/tree/master/examples) benchmark scripts (either for TensorFlow or PyTorch).  

See how the performance compares going from 1 CPU to 2 to 4, etc.  Then if you're seeing no improvement, try creating a [timeline](https://github.com/horovod/horovod/blob/master/docs/timeline.rst) so we can see where the process is slowing down.

**Comment by mariamrahmaani** [comment]
Here is what I did so far for my CPU experiment. 
I setup a completely new system this time with the dockerfile presented in the project: "https://github.com/horovod/horovod/blob/master/Dockerfile.cpu" instead of setting up my system using virtual env. from scratch. 

I just added Bert specific packages and data-set downloaded in this dockerfile.  and some other configs for my proxies. 

so far I ran it with this command line inside docker (not using distributed horovod yet):

horovodrun -np 4 -H localhost:4  python run_glue_horovod.py   --model_type bert   --model_name_or_path ~/bert-checkpoint   --task_name MRPC   --do_train   --do_eval   --do_lower_case   --data_dir ~/GLUE/GLUE/glue_data/MRPC   --max_seq_length 128   --per_gpu_train_batch_size 32   --learning_rate 2e-5   --num_train_epochs 3.0   --output_dir ~/output_dir/MPRC

I see better accuracy results, better results than previous environment with virtual env.  setup in ubuntu:

features from cached file /root/GLUE/GLUE/glue_data/MRPC/cached_dev_bert-checkpoint_128_mrpc
[1,2]<stderr>:11/11/2019 22:30:29 - INFO - __main__ -   ***** Running evaluation  *****
[1,2]<stderr>:11/11/2019 22:30:29 - INFO - __main__ -     Num examples = 408
[1,2]<stderr>:11/11/2019 22:30:29 - INFO - __main__ -     Batch size = 8
Evaluating: 100%|##########| 51/51 [02:01<00:00,  2.39s/it][1,3]<stderr>:
[1,3]<stderr>:11/11/2019 22:32:30 - INFO - __main__ -   ***** Eval results  *****
[1,3]<stderr>:11/11/2019 22:32:30 - INFO - __main__ -     acc = 0.8627450980392157
[1,3]<stderr>:11/11/2019 22:32:30 - INFO - __main__ -     acc_and_f1 = 0.8813725490196078
[1,3]<stderr>:11/11/2019 22:32:30 - INFO - __main__ -     f1 = 0.9
Evaluating: 100%|##########| 51/51 [02:03<00:00,  2.42s/it][1,2]<stderr>:
[1,2]<stderr>:11/11/2019 22:32:32 - INFO - __main__ -   ***** Eval results  *****
[1,2]<stderr>:11/11/2019 22:32:32 - INFO - __main__ -     acc = 0.8627450980392157
[1,2]<stderr>:11/11/2019 22:32:32 - INFO - __main__ -     acc_and_f1 = 0.8813725490196078
[1,2]<stderr>:11/11/2019 22:32:32 - INFO - __main__ -     f1 = 0.9
Evaluating: 100%|##########| 51/51 [02:06<00:00,  2.47s/it][1,1]<stderr>:
[1,1]<stderr>:11/11/2019 22:32:34 - INFO - __main__ -   ***** Eval results  *****
[1,1]<stderr>:11/11/2019 22:32:34 - INFO - __main__ -     acc = 0.8627450980392157
[1,1]<stderr>:11/11/2019 22:32:34 - INFO - __main__ -     acc_and_f1 = 0.8813725490196078
[1,1]<stderr>:11/11/2019 22:32:34 - INFO - __main__ -     f1 = 0.9
Evaluating: 100%|##########| 51/51 [02:07<00:00,  2.49s/it][1,0]<stderr>:
[1,0]<stderr>:11/11/2019 22:32:35 - INFO - __main__ -   ***** Eval results  *****
[1,0]<stderr>:11/11/2019 22:32:35 - INFO - __main__ -     acc = 0.8627450980392157
[1,0]<stderr>:11/11/2019 22:32:35 - INFO - __main__ -     acc_and_f1 = 0.8813725490196078
[1,0]<stderr>:11/11/2019 22:32:35 - INFO - __main__ -     f1 = 0.9

I still don't know which discrepancy between my virtual env. and this dockerfile made this accuracy drop.

Next step is to run the commands in the distributed docker setup and see if the results will be as good as this one.
